{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc858c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(dplyr)\n",
    "library(Matrix)\n",
    "library(Signac)\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(GenomicRanges)\n",
    "library(AnnotationHub)\n",
    "library(patchwork)\n",
    "library(ensembldb)\n",
    "library(biovizBase)\n",
    "library(TFBSTools)\n",
    "library(JASPAR2020)\n",
    "library(Biostrings)\n",
    "library(ggseqlogo)\n",
    "library(universalmotif)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc78acb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rda_path <- \"/data/imindol02/COVID_multiome/COVID_Objects/COVID_filter6.rda\"\n",
    "load(rda_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4511189-6c95-4be6-8210-4615326a27e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- COVID_filter6\n",
    "print(pbmc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8219dd45-8207-4586-8e31-9fe0284ef05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments_file = '/data/imindol02/COVID_multiome/COVID_ATAC/atac_fragments.tsv.gz'\n",
    "\n",
    "counts_data <- pbmc[['peaks']]$counts\n",
    "new_chrom_assay <- CreateChromatinAssay(\n",
    "  counts = counts_data,        # 2단계에서 추출한 counts\n",
    "  sep = c(\":\", \"-\"),\n",
    "  fragments = fragments_file,  # 1단계에서 지정한 원본 경로\n",
    "  min.cells = 10,\n",
    "  min.features = 200\n",
    ")\n",
    "\n",
    "pbmc[['peaks']] <- new_chrom_assay\n",
    "class(pbmc[['peaks']])\n",
    "DefaultAssay(pbmc) <- \"peaks\"\n",
    "\n",
    "frag_objects <- Fragments(pbmc)\n",
    "print(frag_objects[[1]]@path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a187f2be-7370-4119-8937-3d0855e5227d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ah <- AnnotationHub()\n",
    "\n",
    "query(ah, c(\"EnsDb\", \"Homo sapiens\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e8ab918-1eb3-4e6b-9df7-3a2b7288e929",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(timeout = 3600)\n",
    "ensdb_latest <- ah[[\"AH53211\"]]\n",
    "\n",
    "annotations <- GetGRangesFromEnsDb(ensdb = ensdb_latest)\n",
    "seqlevelsStyle(annotations) <- 'UCSC'\n",
    "genome(annotations) <- \"hg38\"\n",
    "Annotation(pbmc) <- annotations\n",
    "print(pbmc[['peaks']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6145772c-25a9-4fee-a779-aa5c7e8a5ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_levels <- c('B','Plasmablast','CD4_T','Treg','MAIT','CD8_T','NKT','NK','CD14_Monocyte','CD16_Monocyte','cDC1','cDC2','pDC')\n",
    "pbmc@active.ident <- factor(x = pbmc@active.ident, levels = my_levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00696d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# B / T subclustering by loading csv\n",
    "\n",
    "full_metadata <- read.csv('/data/imindol02/COVID_multiome/COVID_DataFrames/COVID_subset/updated_metadata_B.csv', row.names = 1)\n",
    "#full_metadata <- read.csv('/data/imindol02/COVID_multiome/COVID_DataFrames/COVID_subset/updated_metadata_T.csv', row.names = 1)\n",
    "\n",
    "pbmc <- AddMetaData(object = pbmc, \n",
    "                          metadata = full_metadata['B_subcluster'],  # 'T_subcluster'\n",
    "                          col.name = 'B_subcluster')  # 'T_subcluster'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "718d7b1c-dbb3-493a-833d-86275ace7e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_metadata <- read.csv('/data/imindol02/COVID_multiome/COVID_DataFrames/COVID_subset/updated_metadata_celltype_abs.csv', row.names = 1)\n",
    "\n",
    "pbmc <- AddMetaData(object = pbmc, \n",
    "                          metadata = full_metadata['cell_type_abstract'],\n",
    "                          col.name = 'celltype_abstract')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1294fb98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bc020ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(species = 9606, all_versions = FALSE) # 9606=Human, 10090=Mouse\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70fb4a0b-5987-4a71-bb42-5d8a5fe41680",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- AddMotifs(\n",
    "  object = pbmc,\n",
    "  genome = BSgenome.Hsapiens.UCSC.hg38,\n",
    "  pfm = pfm,\n",
    "  assay = \"peaks\" \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9badfd3c-7a42-41d8-9ba4-03dc3bbe6df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- RegionStats(\n",
    "  object = pbmc,\n",
    "  genome = BSgenome.Hsapiens.UCSC.hg38,\n",
    "  assay = \"peaks\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e90d508a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84f0bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "B_subset <- subset(x = pbmc, subset = big_anno %in% c(\"B\", \"Plasmablast\"))\n",
    "\n",
    "Idents(B_subset) <- \"B_subcluster\"\n",
    "\n",
    "my_levels <- c('Naive B cell group 1','Naive B cell group 2','Activated Naive B cell','Memory B cell','Transitional B cell','atypical Memory B cell','atypical B cell','Plasmablast')\n",
    "B_subset@active.ident <- factor(x = B_subset@active.ident, levels = my_levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2af553a5-14d6-48c6-8638-33db71cb6185",
   "metadata": {},
   "outputs": [],
   "source": [
    "T_subset <- subset(x = pbmc, subset = big_anno %in% c(\"CD4_T\", \"CD8_T\", \"MAIT\", \"Treg\"))\n",
    "\n",
    "Idents(T_subset) <- \"T_subcluster\"\n",
    "\n",
    "my_levels <- c('Naive T','Naive CD4','Naive CD8','CM CD4','EM CD4','Treg','CTL','MAIT')\n",
    "T_subset@active.ident <- factor(x = T_subset@active.ident, levels = my_levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a71bbe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "337e7a6c-7c49-46ac-8a26-ca0d9eef4e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "atBC_DAR <- FindMarkers(\n",
    "  object = B_subset,\n",
    "  ident.1 = \"atypical B cell\", # change to 'atypical Memory B cell' / change to atMBC_DAR\n",
    "  group.by = \"B_subcluster\",\n",
    "  only.pos = TRUE,\n",
    "  test.use = 'LR'\n",
    ")\n",
    "\n",
    "atBC_DAR_peak <- rownames(atBC_DAR[atBC_DAR$p_val < 0.05, ])\n",
    "\n",
    "atBC_motif <- FindMotifs(\n",
    "  object = B_subset,\n",
    "  features = atBC_DAR_peak\n",
    ")\n",
    "\n",
    "atBC_motif_df <- atBC_motif\n",
    "\n",
    "plot_data <- atBC_motif_df %>%\n",
    "  arrange(pvalue) %>%\n",
    "  mutate(log_p = -log10(pvalue))\n",
    "\n",
    "label_data <- plot_data %>%\n",
    "  head(30)\n",
    "\n",
    "p <- ggplot(plot_data, aes(x = fold.enrichment, y = log_p)) +\n",
    "  \n",
    "  geom_point(aes(color = log_p), size = 1.5, alpha = 1.0) +\n",
    "  scale_color_gradient(low = \"grey\", high = \"#E41A1C\") +\n",
    "  \n",
    "  geom_text_repel(\n",
    "    data = label_data,\n",
    "    aes(label = motif.name),\n",
    "    box.padding = 1.0,\n",
    "    point.padding = 1.0,\n",
    "    max.overlaps = Inf,\n",
    "    #min.segment.length = 0,\n",
    "    segment.color = \"grey80\",\n",
    "    segment.size = 0.3,\n",
    "    size = 3.5,\n",
    "    fontface = \"plain\", # bold\n",
    "    color = \"black\"\n",
    "  ) +\n",
    "  \n",
    "  theme_classic() +\n",
    "  labs(\n",
    "    title = \"Enriched Motifs\",\n",
    "    subtitle = \"Labels: top 30 significant motifs\",\n",
    "    x = \"Fold Enrichment\",\n",
    "    y = \"-log10(P-value)\",\n",
    "    color = \"Significance\"\n",
    "  ) +\n",
    "  theme(\n",
    "    axis.title = element_text(size = 12, face = \"bold\"),\n",
    "    axis.text = element_text(size = 10),\n",
    "    plot.title = element_text(hjust = 0.5, face = \"bold\")\n",
    "  )\n",
    "\n",
    "\n",
    "for (i in seq_len(nrow(label_data))) {\n",
    "  row_data <- label_data[i, ]\n",
    "  motif_id <- row_data$motif\n",
    "  q <- MotifPlot(\n",
    "      object = B_subset,\n",
    "      motifs = motif_id,\n",
    "      assay = 'peaks'\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f355bac9-36c3-4470-a581-7c1ac63135ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAIT_DAR <- FindMarkers(\n",
    "  object = T_subset,\n",
    "  ident.1 = \"MAIT\",\n",
    "  group.by = \"T_subcluster\",\n",
    "  only.pos = TRUE,\n",
    "  test.use = 'LR'\n",
    ")\n",
    "\n",
    "MAIT_DAR_peak <- rownames(MAIT_DAR[MAIT_DAR$p_val < 0.05, ])\n",
    "\n",
    "MAIT_motif <- FindMotifs(\n",
    "  object = T_subset,\n",
    "  features = MAIT_DAR_peak\n",
    ")\n",
    "\n",
    "MAIT_motif_df <- MAIT_motif\n",
    "\n",
    "plot_data <- MAIT_motif_df %>%\n",
    "  arrange(pvalue) %>%\n",
    "  mutate(log_p = -log10(pvalue))\n",
    "\n",
    "label_data <- plot_data %>%\n",
    "  head(30)\n",
    "  \n",
    "p <- ggplot(plot_data, aes(x = fold.enrichment, y = log_p)) +\n",
    "  \n",
    "  geom_point(aes(color = log_p), size = 1.5, alpha = 1.0) +\n",
    "  scale_color_gradient(low = \"grey\", high = \"#E41A1C\") +\n",
    "  \n",
    "  geom_text_repel(\n",
    "    data = label_data,\n",
    "    aes(label = motif.name),\n",
    "    box.padding = 1.0,\n",
    "    point.padding = 1.0,\n",
    "    max.overlaps = Inf,\n",
    "    #min.segment.length = 0,\n",
    "    segment.color = \"grey80\",\n",
    "    segment.size = 0.3,\n",
    "    size = 3.5,\n",
    "    fontface = \"plain\", # bold\n",
    "    color = \"black\"\n",
    "  ) +\n",
    "  \n",
    "  theme_classic() +\n",
    "  labs(\n",
    "    title = \"Enriched Motifs\",\n",
    "    subtitle = \"Labels: top 30 significant motifs\",\n",
    "    x = \"Fold Enrichment\",\n",
    "    y = \"-log10(P-value)\",\n",
    "    color = \"Significance\"\n",
    "  ) +\n",
    "  theme(\n",
    "    axis.title = element_text(size = 12, face = \"bold\"),\n",
    "    axis.text = element_text(size = 10),\n",
    "    plot.title = element_text(hjust = 0.5, face = \"bold\")\n",
    "  )\n",
    "\n",
    "\n",
    "for (i in seq_len(nrow(label_data))) {  \n",
    "  row_data <- label_data[i, ]\n",
    "  motif_id <- row_data$motif\n",
    "  q <- MotifPlot(\n",
    "      object = T_subset,\n",
    "      motifs = motif_id,\n",
    "      assay = 'peaks'\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bbfbb06",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec5304e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Coverageplot\n",
    "\n",
    "roi <- StringToGRanges(c(\"chr19-41330323-41353922\"))\n",
    "\n",
    "CoveragePlot(\n",
    "  object = B_subset,\n",
    "  region = \"TGFB1\",\n",
    "  region.highlight = roi,\n",
    "  group.by = \"B_subcluster\",\n",
    "  extend.upstream = 1500,\n",
    "  extend.downstream = 500,\n",
    "  width = 25, height = 15\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594dadc6-ed2c-4d19-8b68-438edc02ea75",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_genes <- c(\"TGFB1\")\n",
    "\n",
    "B_subset <- LinkPeaks(\n",
    "  object = B_subset,\n",
    "  peak.assay = \"peaks\",\n",
    "  expression.assay = \"RNA\", # or \"SCT\"\n",
    "  genes.use = target_genes,\n",
    "  distance = 500000,\n",
    "  pvalue_cutoff = 0.05,\n",
    "  score_cutoff = 0.05\n",
    ")\n",
    "\n",
    "links_df <- as.data.frame(Links(B_subset))\n",
    "\n",
    "linked_peaks <- links_df %>%\n",
    "  filter(gene == target_gene & score > 0.05) %>%\n",
    "  pull(peak)\n",
    "\n",
    "enriched_motifs <- FindMotifs(\n",
    "  object = B_subset,\n",
    "  features = linked_peaks\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2becc279-e7c1-4b50-b860-30af290896c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "TGFB1_linked_motifs <- enriched_motifs %>%\n",
    "  filter(observed == 1) %>%\n",
    "  pull(motif.name)\n",
    "length(TGFB1_linked_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0773e8c3-9b38-47a8-a0c1-ecad7d042a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "atBC_linked_motifs <- atBC_motif %>%\n",
    "  filter(pvalue < 0.05) %>%\n",
    "  pull(motif.name)\n",
    "length(atBC_linked_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6a7c58f-f9f5-4b6b-a335-861011b89978",
   "metadata": {},
   "outputs": [],
   "source": [
    "MOTIF_RESULT <- intersect(TGFB1_linked_motifs, atBC_linked_motifs)\n",
    "length(MOTIF_RESULT)\n",
    "print(MOTIF_RESULT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "274ff513-d390-4a76-9af6-8edcb2f37c11",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d919889a-20f6-4a29-a1b6-99466af0e15f",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_annotation <- Annotation(B_subset)\n",
    "target_genes <- c(\"TGFB1\") \n",
    "subset_annotation <- total_annotation[total_annotation$gene_name %in% target_genes]\n",
    "B_subset_forPlot <- B_subset\n",
    "Annotation(B_subset_forPlot) <- subset_annotation\n",
    "\n",
    "p <- CoveragePlot(\n",
    "  object = B_subset_forPlot,\n",
    "  region = \"TGFB1\",\n",
    "  features = \"TGFB1\",\n",
    "  expression.assay = \"RNA\",\n",
    "  links = TRUE,\n",
    "  extend.upstream = 20000,\n",
    "  extend.downstream = 300000\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12580a28-51d9-4af2-b28f-b520b99ea93b",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi <- StringToGRanges(c(\"chr19-41570170-41570225\"))\n",
    "\n",
    "p <- CoveragePlot(\n",
    "  object = B_subset_forPlot,\n",
    "  region = \"chr19-41569966-41570457\",\n",
    "  region.highlight = roi,\n",
    "  feature = \"TGFB1\",\n",
    "  expression.assay = \"RNA\",\n",
    "  width = 25, height = 15,\n",
    "  annotation = FALSE,\n",
    "  peaks = FALSE,\n",
    "  links = FALSE\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R_atac",
   "language": "R",
   "name": "r_atac"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
